### PR TITLE
Automatically refresh SSL certs before they expire, part 2

### PR DESCRIPTION
**Pull request by jellybeanfiend**
For issue #94, Follow up of https://github.com/GoogleCloudPlatform/cloudsql-proxy/pull/142 which was reverted due to some unaddressed comments.

### CODE REVIEW

File: proxy/proxy/client_test.go [hunk 1]
```diff
@@ -183,3 +192,54 @@ func TestMaximumConnectionsCount(t *testing.T) {
...
+	}
+
+	c.cacheL.Lock()
+	cached, ok := c.cfgCache[instance]
+	c.cacheL.Unlock()
+	if !ok {
+		t.Error("expected instance to be cached")
+	}
+	waitTil := time.After(timeToExpire + (10 * time.Millisecond))
+loop:
```
Comment (Carrotman42) [inline proxy/proxy/client_test.go]: Maybe this came from a previous review comment of mine, but: this loop doesn't do anything different than simply executing this sleep:
   
    time.Sleep(timeToExpire + 10*time.Millisecond)

I somewhat recall asking for a sleep+retry loop. If that's what you were attempting to implement here, you should be checking the cfgCache multiple times and sleep+retrying if the config you were looking for was not found/refreshed. That will reduce flakiness by giving the background goroutines some time to process while still reducing the amount of time the test takes in the common case.

File: proxy/proxy/client.go [hunk 2]
```diff
@@ -195,16 +207,33 @@ func (c *Client) refreshCfg(instance string) (addr string, cfg *tls.Config, err
 		Certificates: []tls.Certificate{mycert},
 		RootCAs:      certs,
 	}
+
+	// Refresh cert 5 minutes before it expires.
+	timeToRefresh := cfg.Certificates[0].Leaf.NotAfter.Sub(time.Now()) - refreshCertBuffer
```
Comment (Carrotman42) [inline proxy/proxy/client.go]: nit: please use `mycert` here instead of `cfg.Certificates[0]`

File: proxy/proxy/client.go [hunk 3]
```diff
@@ -150,31 +156,37 @@ func (c *Client) handleConn(conn Conn) {
...
-	}
-
 	defer func() {
-		c.cfgCache[instance] = cacheEntry{
-			lastRefreshed: time.Now(),
-
-			err:  err,
-			addr: addr,
-			cfg:  cfg,
+		if err == nil || (oldok && isExpired(old.cfg)) {
```
Comment (Carrotman42) [inline proxy/proxy/client.go]: Still not all the cases :) You still want to update the cache in the case that the first call to refreshCfg for an instance returned an error: in that case, `oldok` would be false and `err` would be non-nil.

**Comment by kurtisvg** [comment]
@jellybeanfiend Are you planning on continuing work on this PR?

**Comment by hfwang** [comment]
jellybeanfiend has mentioned they will not be continuing to work on this. I'll find someone internally to pick this back up.

**Pull request closed without merging**